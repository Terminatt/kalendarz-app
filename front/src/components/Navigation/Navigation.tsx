import { RootState } from '@store/index';
import { Menu } from 'antd';
import React, { useCallback, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { MailOutlined } from '@ant-design/icons';
import CustomMenu from '@components/CustomMenu/CustomMenu';
import { useNavigate, useLocation } from 'react-router-dom';

import './Navigation.less';
import useAdmin from '@hooks/useAdmin';
import useLogged from '@hooks/useLogged';

const Navigation: React.FC = () => {
    const [openKeys, setOpenKeys] = useState<string[]>([]);
    const navigate = useNavigate();
    const location = useLocation();

    const isAdmin = useAdmin();
    const isLogged = useLogged();
    const selectedKeys = location.pathname.split('/');

    const onSelect = useCallback((info: {key: string, keyPath: string[]}) => {
        const { keyPath } = info;
        // the keypath is constructed from bottom to top: Menu.Item key -> SubMenu key
        const url = keyPath.reverse().join('/');
        navigate(url);
    }, []);

    const onOpen = useCallback((keys: string[]) => {
        setOpenKeys(keys);
    }, []);

    useEffect(() => {
        const isIncluded = openKeys.includes(selectedKeys[0]);
        if (isIncluded) {
            return;
        }

        const tmp = [...openKeys];
        tmp.push(selectedKeys[1]);
        setOpenKeys(tmp);
    }, [location.pathname]);

    return isLogged ? (
        <nav>
            <CustomMenu
                onOpenChange={onOpen}
                selectedKeys={selectedKeys}
                openKeys={openKeys}
                onSelect={onSelect}
                mode="inline"
            >
                <Menu.SubMenu key="user-zone" icon={<MailOutlined />} title="Strefa Użytkownika">
                    <Menu.Item key="my-reservations">Moje Rezerwacje</Menu.Item>
                    <Menu.Item key="my-account">Moje Konto</Menu.Item>
                </Menu.SubMenu>
                {isAdmin ? (
                    <Menu.SubMenu key="admin-zone" icon={<MailOutlined />} title="Strefa Administratora">
                        <Menu.Item key="users">Użytkownicy</Menu.Item>
                        <Menu.Item key="rooms">Pokoje</Menu.Item>
                        <Menu.Item key="room-types">Typy pokojów</Menu.Item>
                    </Menu.SubMenu>
                ) : null}
            </CustomMenu>
        </nav>
    ) : (
        <div>Nawigacja jest dostępna po zalogowaniu</div>
    );
};

export default Navigation;
