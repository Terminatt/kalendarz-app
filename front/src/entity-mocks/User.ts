import { Group, User } from '@store/user/types';
import dayjs from 'dayjs';

export const userMock: User = {
    id: 1,
    firstName: 'Test',
    lastName: 'User',
    created: dayjs().toISOString(),
    email: 'test@gmail.com',
    username: 'Test User',
    groups: Group.REGULAR_USER,
};

export const adminUserMock: User = {
    id: 2,
    firstName: 'Test',
    lastName: 'User',
    created: dayjs().toISOString(),
    email: 'test@gmail.com',
    username: 'Test User',
    groups: Group.ADMIN,
};
