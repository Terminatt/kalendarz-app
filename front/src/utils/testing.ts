export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
