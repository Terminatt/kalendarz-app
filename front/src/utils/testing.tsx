import React, { ComponentType } from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { RootState, rootReducer } from '@store/index';

export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
export function isTestingEnv(): boolean {
    return process.env.JEST_WORKER_ID !== undefined || process.env.NODE_ENV === 'test';
}

export interface CustomOptions extends RenderOptions {
    preloadedState: RootState;
}

function render(
    ui: React.ReactElement,
    options?: CustomOptions,
) {
    const store = configureStore({ reducer: rootReducer, preloadedState: options?.preloadedState });
    function Wrapper({ children }: {children: React.ReactNode}) {
        return <Provider store={store}>{children}</Provider>;
    }

    return { element: rtlRender(ui, { wrapper: Wrapper as ComponentType, ...options }), store };
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
