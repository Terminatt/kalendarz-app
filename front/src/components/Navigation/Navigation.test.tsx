

import { userMock, adminUserMock } from '@entity-mocks/User';
import { store } from '@store/index';
import { render } from '@utils/testing';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import Navigation from './Navigation';

describe('Navigation Component', () => {
    beforeEach(() => {})

    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <MemoryRouter>
                        <Navigation />
                    </MemoryRouter>
                </Provider>
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('renders text if user in not logged it', () => {
        const { element } = render(
            <MemoryRouter>
                <Navigation  />
            </MemoryRouter>
        );

        expect(element.getByText('Nawigacja jest dostępna po zalogowaniu')).not.toBeNull();
    });

    it('it shows user zone when logged in', async () => {
        
        const { element } = render(
            <MemoryRouter>
                <Navigation  />
            </MemoryRouter>, 
            {
                preloadedState: {user: {currentUser: userMock}}
            }
        );
        
        expect(element.getByText('Strefa Użytkownika')).not.toBeNull();
    });

    it('it does not shows admin zone when logged in as regular user', async () => {
        
        const { element } = render(
            <MemoryRouter>
                <Navigation  />
            </MemoryRouter>, 
            {
                preloadedState: {user: {currentUser: userMock}}
            }
        );
        
        expect(await element.queryByText('Strefa Administratora')).toBeNull();
    });

    it('it shows user zone when logged in as admin user', async () => {
        
        const { element } = render(
            <MemoryRouter>
                <Navigation  />
            </MemoryRouter>, 
            {
                preloadedState: {user: {currentUser: adminUserMock}}
            }
        );
        
        expect(await element.findByText('Strefa Użytkownika')).not.toBeNull();
    });

    it('it shows admin zone when logged in as admin user', async () => {
        
        const { element } = render(
            <MemoryRouter>
                <Navigation  />
            </MemoryRouter>, 
            {
                preloadedState: {user: {currentUser: adminUserMock}}
            }
        );
        
        expect(await element.findByText('Strefa Administratora')).not.toBeNull();
    });
});
