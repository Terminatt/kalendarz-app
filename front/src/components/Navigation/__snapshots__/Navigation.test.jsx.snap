// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Component matches the snapshot 1`] = `
<div>
  Nawigacja jest dostÄ™pna po zalogowaniu
</div>
`;
